lib/actions-oclif/build.ts

```diff
@@ -0,0 +1,271 @@
+				app,
+				arch: options.arch!,
+				deviceType: options.deviceType!,
+				buildEmulated: options.emulated,
+				buildOpts,
+			});
+		} catch (err) {
+			logger.logError('Build failed.');
+			throw err;
+		}
+
+		logger.outputDeferredMessages();
+		logger.logSuccess('Build succeeded!');
+	}
+
+	protected translateParams(params: ArgsDef, options: FlagsDef) {
+		// Copy flags to those expected by other modules
+		options.arg = options.buildArg;
+		delete options.buildArg;
+		options['image-list'] = options['cache-from'];
+		delete options['cache-from'];
+
+		// `build` accepts `[source]` as a parameter, but compose expects it
+		// as an option. swap them here
+		if (options.source == null) {
+			options.source = params.source;
+		}
+		delete params.source;
+	}
+
+	protected async validateOptions(opts: FlagsDef, sdk: BalenaSDK) {
+		// Validate option combinations
+		if (
+			(opts.application == null &&
+				(opts.arch == null || opts.deviceType == null)) ||
+			(opts.application != null &&
+				(opts.arch != null || opts.deviceType != null))
+		) {
+			const { ExpectedError } = await import('../errors');
+			throw new ExpectedError(
+				'You must specify either an application or an arch/deviceType pair to build for',
+			);
+		}
+
+		// Validate project directory
+		const { validateProjectDirectory } = await import('../utils/compose_ts');
+		const { dockerfilePath, registrySecrets } = await validateProjectDirectory(
+			sdk,
+			{
+				dockerfilePath: opts.dockerfile,
+				noParentCheck: opts['noparent-check'] || false,
+				projectPath: opts.source || '.',
+				registrySecretsPath: opts['registry-secrets'],
+			},
+		);
+
+		opts.dockerfile = dockerfilePath;
+		opts['registry-secrets'] = registrySecrets;
+	}
+
+	protected async getAppAndResolveArch(opts: FlagsDef) {
+		if (opts.application) {
+			const { getAppWithArch } = await import('../utils/helpers');
+			const app = await getAppWithArch(opts.application);
+			opts.arch = app.arch;
+			opts.deviceType = app.device_type;
+			return app;
+		}
+	}
+
+	protected async prepareBuild(options: FlagsDef) {
+		const { getDocker, generateBuildOpts } = await import('../utils/docker');
+		const [docker, buildOpts, composeOpts] = await Promise.all([
+			getDocker(options),
+			generateBuildOpts(options),
+			compose.generateOpts(options),
+		]);
+		return {
+			docker,
+			buildOpts,
+			composeOpts,
+		};
+	}
+
+	/**
+	 * Opts must be an object with the following keys:
+	 *   app: the app this build is for (optional)
+	 *   arch: the architecture to build for
+	 *   deviceType: the device type to build for
+	 *   buildEmulated
+	 *   buildOpts: arguments to forward to docker build command
+	 *
+	 * @param {import('docker-toolbelt')} docker
+	 * @param {import('../utils/logger')} logger
+	 * @param {import('../utils/compose-types').ComposeOpts} composeOpts
+	 * @param {any} opts
+	 */
+	protected async buildProject(
+		docker: DockerToolbelt,
+		logger: Logger,
```

Comment:
`Logger` type import required because of this.

Reply:
The alternative to a module level `import type` is:
```suggestion
		logger: import('../utils/logger'),
```

Reply:
Thanks.  Perhaps we can grab the Logger type from the base Command class (in a way that doesn't affect performance if logger is not used)?

Reply:
This import is purely in typings so will only exist for the typescript compiler - at runtime none of the type info will exist (even if it's a non-type only module level import it'll probably still be elided if it's only used for types, but I prefer the unambiguous ways of doing it because it avoids accidents)